shared/readline: Disable auto-indent if space/tab is first on line.

dpgeorge opened the pull request:
So that text can be pasted to the normal REPL, like CPython.

Auto-indent still works.

Addresses issue #7925.

dpgeorge committed:
shared/readline: Disable auto-indent if space/tab follows added indent.

Auto-indent still works as the default behaviour, but it is now undone and
disabled if there is a space/tab immediately after an automatically-added
indent.  This makes the REPL behaviour closer to CPython, and in particular
allows text to be pasted at the normal REPL.

Addresses issue #7925.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/cmdline: Add test for REPL auto-indent.

Signed-off-by: Damien George <damien@micropython.org>

robert-hh commented:
That works at the REPL prompt fine if spaces are used for indentation. When using tabs, the behavior is kind of strange. Each TAB causes a tab completion list to be shown, but the end, the right lines are inserted. A small difference to the space-indentation is, that at the last line the cursor is indented.
text used to the test:
```
if True:
	print("Something")
	print("else")
```
Result at the screen:
```
>>> if True:
...     
Pin             SDCard          mimxrt          sys
os              bdev            vfs             sdcard
...     print("Something")
...     
Pin             SDCard          mimxrt          sys
os              bdev            vfs             sdcard
...     print("else")
... 
Something
else
>>> 
```

dpgeorge commented:
> That works at the REPL prompt fine if spaces are used for indentation
> ...
> Tested with a Teensy 4.0, master branch

I'm not sure what you are saying.  Are you saying that without this PR pasting already works OK at the normal REPL?

Using the above test code on master unix port (ie without this PR) I get:
```
MicroPython v1.18-330-gd242a9b7f on 2022-04-13; linux [GCC 11.2.0] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> if True:
...             print("Something")
...                     print("else")
...                     
...                     
... 
Traceback (most recent call last):
  File "<stdin>", line 3
IndentationError: unexpected indent
>>> 
```
and with this PR I get:
```
MicroPython v1.18-332-gb78b4ee7d on 2022-04-14; linux [GCC 11.2.0] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> if True:
...     print("Something")
...     print("else")
... 
Something
else
```

robert-hh commented:
Sorry for  a possible confusion. I'm only talking about the behavior with that PR and the difference between pasting code either indented with spaces or indented with tabs. Using linux micropython, I get the same behavior as you. Pasting the few lines form above, which are indented with tabs,  at the REPL of a board, like Pyboard, ESP32, RPi Pico or Teensy, I get this picture shown above. Every line starting with a tab character causes the output of suggested completions.
The code is entered fine finally.

dpgeorge committed:
shared/readline: Make tab insert an indent when it follows whitespace.

Entering tab at the REPL will now make it insert an indent (4 spaces) in
the following cases:
- after any whitespace on a line
- at the start of a line that is not the first line

This changes the existing behaviour where a tab would insert an indent only
if there were no matches in the auto-complete search, and it was the start
of the line.  This means, if there were any symbols in the global
namespace, tab could never be used to indent.

Note that entering tab at the start of the first line will still do
auto-completion, but will now do nothing if there are no symbols in the
global namespace, which is more consistent than before.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
> Pasting the few lines form above, which are indented with tabs, at the REPL of a board, like Pyboard, ESP32, RPi Pico or Teensy, I get this picture shown above. Every line starting with a tab character causes the output of suggested completions.

You are correct.  I only tested on unix normal, not other builds/ports which change the behaviour when a global symbol exists.

I have fixed this case and as part of that I think made a good improvement to the use of tab: it can now be used to insert an indent reliably, regardless of whether any symbols exist in the global name space.

From the commit message:

Entering tab at the REPL will now make it insert an indent (4 spaces) in the following cases:
- after any whitespace on a line
- at the start of a line that is not the first line
    
This changes the existing behaviour where a tab would insert an indent only if there were no matches in the auto-complete search, and it was the start of the line.  This means, if there were any symbols in the global namespace, tab could never be used to indent.
    
Note that entering tab at the start of the first line will still do auto-completion, but will now do nothing if there are no symbols in the global namespace, which is more consistent than before.
